tests/test_cli_compile.py

```diff
@@ -2702,22 +2707,240 @@ def test_all_extras_fail_with_extra(fake_dists, runner, make_module, fname, cont
     assert exp in out.stderr
 
 
+def _mock_resolver_cls(monkeypatch: MonkeyPatch) -> MagicMock:
+    obj = MagicMock()
+    obj.resolve = MagicMock(return_value=set())
+    obj.resolve_hashes = MagicMock(return_value=dict())
+    cls = MagicMock(return_value=obj)
+
+    monkeypatch.setattr("piptools.scripts.compile.BacktrackingResolver", cls)
+    monkeypatch.setattr("piptools.scripts.compile.LegacyResolver", cls)
+
+    return cls
+
+
+def _mock_build_project_metadata(monkeypatch: MonkeyPatch) -> MagicMock:
+    func = MagicMock(
+        return_value=ProjectMetadata(
+            extras=("e",),
+            requirements=(
+                install_req_from_line("rdep0"),
+                install_req_from_line("rdep1; extra=='e'"),
+            ),
+            build_requirements=(install_req_from_line("bdep0"),),
+        )
+    )
+
+    monkeypatch.setattr("piptools.scripts.compile.build_project_metadata", func)
+
+    return func
+
+
+@backtracking_resolver_only
+@pytest.mark.network
+def test_all_extras_and_all_build_deps(
+    fake_dists_with_build_deps,
+    runner,
+    tmp_path,
+    monkeypatch,
+    current_resolver,
+):
+    """
+    Test that trying to lock all dependencies gives the expected output.
+    """
+    src_pkg_path = pathlib.Path(PACKAGES_PATH) / "small_fake_with_build_deps"
+    # When used as argument to the runner it is not passed to pip
+    monkeypatch.setenv("PIP_FIND_LINKS", fake_dists_with_build_deps)
+
+    with runner.isolated_filesystem(tmp_path) as tmp_pkg_path:
+        shutil.copytree(src_pkg_path, tmp_pkg_path, dirs_exist_ok=True)
+        out = runner.invoke(
+            cli,
+            [
+                "--allow-unsafe",
+                "--output-file",
+                "-",
+                "--quiet",
+                "--no-emit-options",
+                "--no-header",
+                "--all-extras",
+                "--all-build-deps",
+            ],
+        )
+
+    assert out.exit_code == 0
```

Comment:
It has been annoying me a bit that when `assert`s such as this fail, we get little information about the cause without first instrumenting the code. What do you think about adding a helper function like the one below? In a separate PR of course.

```python
def assert_ok(
    result: Result, stdout: str | None = None, stderr: str | None = None
) -> None:
    # You may need to disable --numprocesses=auto to see these prints in your IDE

    actual = {"exit_code": result.exit_code}
    expected = {"exit_code": 0}

    if stdout is not None:
        actual["stdout"] = result.stdout
        expected["stdout"] = stdout
    elif result.stdout:
        print("stdout:")
        print(result.stdout)

    if stderr is not None:
        actual["stderr"] = result.stderr
        expected["stderr"] = stderr
    elif result.stderr:
        print("stderr:")
        print(result.stderr)

    if result.exc_info:
        print("exc_info:")
        traceback.print_exception(*result.exc_info, file=sys.stdout)

    assert actual == expected
```

Reply:
Would be used e.g. like

```diff
-    assert out.exit_code == 0
-    assert out.stdout == dedent(
-        """\
-        small-fake-a==0.1
-            # via
-            #   -c constraints.txt
-            #   -r requirements.in
-        """
+    assert_ok(
+        result=out,
+        stdout=dedent(
+            """\
+            small-fake-a==0.1
+                # via
+                #   -c constraints.txt
+                #   -r requirements.in
+            """
+        ),
     )
```

Reply:
I wouldn't add a helper, but maybe add the second message argument to the assert itself? Also, put things you're comparing into named vars — they'll show up in the error output with `--ll`.

Reply:
Thanks for the suggestions, I will explore what I can do with `--ll`.